Add support for IPv6 Privacy Extension RFC 8981

jukkar opened the pull request:
This one adds support for IPv6 privacy extension defined in https://www.rfc-editor.org/rfc/rfc8981

The IPv6 Privacy Extension will utilize SLAAC so that the IID (Interface Identifier) part of IPv6 address we generate is randomized. This enhances privacy and makes it more difficult to track the IPv6 address user. The use of IPv6 PE is mostly useful to client connections as typically servers need a well known IP address to connect to.

In order the deprecated IPv6 addresses to be removed from the system, some co-operation from application is needed. If we remove IP address that is used by an application, then the connection will not work properly any more. That is why I added IP address reference counting so that if there is a socket using the deprecated IP address, the address is not removed.
The application needs to listen IPv6 address deprecation events, and close the socket that is using the deprecated address. The echo-client application is changed to support this.

Example of the echo-client run:

When the device starts, the IPv6 PE engine is initialized with preferred and valid IP address timeout values.
```
[00:00:00.000,000] <dbg> net_ipv6_pe: net_ipv6_pe_init: (main): pe enabled prefer temporary lifetime 358 sec
*** Booting Zephyr OS build v3.6.0-3080-g3f56999ce23d ***
[00:00:00.000,000] <inf> net_config: Initializing network
[00:00:00.110,000] <inf> net_config: IPv6 address: 2001:db8::1
[00:00:00.110,000] <inf> net_config: IPv6 address: 2001:db8::1
[00:00:00.110,000] <inf> net_echo_client_sample: Run echo client
[00:00:00.110,000] <inf> net_echo_client_sample: Network connected
[00:00:00.120,000] <dbg> net_ipv6_pe: net_ipv6_pe_start: (rx_q[0]): Starting PE process for prefix 2001:db8::/64
[00:00:00.120,000] <dbg> net_ipv6_pe: net_ipv6_pe_start: (rx_q[0]): Lifetime 358 desync 80 timeout 160 preferred 240 valid 360
[00:00:00.120,000] <dbg> net_ipv6_pe: net_ipv6_pe_start: (rx_q[0]): Starting DAD for 2001:db8::5cc:f0bc:93f1:a0e5 iface 1
[00:00:00.120,000] <dbg> net_ipv6_pe: net_ipv6_pe_start: (rx_q[0]): Next check for temp addresses in 160 seconds
[00:00:00.120,000] <inf> net_echo_client_sample: Temporary IPv6 address 2001:db8::5cc:f0bc:93f1:a0e5 added
[00:00:00.230,000] <inf> net_config: IPv6 address: 2001:db8::5cc:f0bc:93f1:a0e5
[00:00:00.230,000] <inf> net_config: IPv6 address: 2001:db8::5cc:f0bc:93f1:a0e5
[00:00:01.130,000] <inf> net_echo_client_sample: Starting...
```
If we receive IPv6 router advertisements that advertise an IPv6 prefix, then a temporary IPv6 address is generated (shown above).
When the echo-client starts, it starts to use the temporary address because the connected socket is marked to prefer temporary addresses.
```
$ net ipv6
...

IPv6 addresses for interface 1 (0x808c854) (Ethernet)
====================================================
Type            State           Lifetime (sec)  Ref     Address
autoconf        preferred             infinite  1       fe80::5eff:fe00:5331/128
manual          preferred             infinite  1       2001:db8::1/64
autoconf        preferred           3239182327  1       2001:db8::5eff:fe00:5331/64
autoconf        preferred                  335  2       2001:db8::5cc:f0bc:93f1:a0e5/64 (temporary)
```
Note that ref count for the temporary address. The second ref count is from the open socket in echo-client.

Then some data is transferred until the address deprecation timeout kicks in.
```
[00:01:01.200,000] <inf> net_echo_client_sample: IPv6 TCP: Exchanged 1000 packets
[00:02:01.200,000] <inf> net_echo_client_sample: IPv6 TCP: Exchanged 2000 packets
[00:02:40.130,000] <dbg> net_ipv6_pe: renewal_cb: (sysworkq): IPv6 address 2001:db8::5cc:f0bc:93f1:a0e5 is deprecated
[00:02:40.130,000] <dbg> net_ipv6_pe: net_ipv6_pe_start: (sysworkq): Starting PE process for prefix 2001:db8::/64
[00:02:40.130,000] <dbg> net_ipv6_pe: net_ipv6_pe_start: (sysworkq): Lifetime 358 desync 41 timeout 119 preferred 160 valid 360
[00:02:40.130,000] <dbg> net_ipv6_pe: net_ipv6_pe_start: (sysworkq): Starting DAD for 2001:db8::7b60:4db3:38a2:bb28 iface 1
[00:02:40.130,000] <dbg> net_ipv6_pe: net_ipv6_pe_start: (sysworkq): Next check for temp addresses in 119 seconds
[00:02:40.130,000] <inf> net_echo_client_sample: Temporary IPv6 address 2001:db8::7b60:4db3:38a2:bb28 added
[00:02:40.240,000] <inf> net_config: IPv6 address: 2001:db8::7b60:4db3:38a2:bb28
[00:02:40.640,000] <inf> net_echo_client_sample: IPv6 address 2001:db8::5cc:f0bc:93f1:a0e5 deprecated
[00:02:40.640,000] <dbg> net_echo_client_sample: wait: Received restart event.
[00:02:40.640,000] <inf> net_echo_client_sample: Stopping...
[00:02:40.640,000] <inf> net_echo_client_sample: Starting...
```
We notice the IPv6 address deprecation and because our TCP socket was bound to that address, we need to close the socket and then re-create it again. Strictly speaking, application could ignore all this in which case the address would not be removed, the connectivity via the deprecated IPv6 address should work fine. It is just that the deprecated addresses are not considered when selecting a source address so all sockets created after the deprecation will use some other address.

After the restart, the socket is using the new source IPv6 address as seen below
```
$ net ipv6
...

IPv6 addresses for interface 1 (0x808c854) (Ethernet)
====================================================
Type            State           Lifetime (sec)  Ref     Address
autoconf        preferred             infinite  1       fe80::5eff:fe00:5331/128
manual          preferred             infinite  1       2001:db8::1/64
autoconf        preferred           3239182306  1       2001:db8::5eff:fe00:5331/64
autoconf        deprecated                 182  1       2001:db8::5cc:f0bc:93f1:a0e5/64 (temporary)
autoconf        preferred                  342  2       2001:db8::7b60:4db3:38a2:bb28/64 (temporary)
```
The deprecated IPv6 addresses are eventually removed when they expire. In the above example, after 182 seconds.

jukkar committed:
net: ipv6: Initial support for privacy extension

This creates support for IPv6 privacy extensions which is described
in RFC 8981. This will also add API that can be used to add IPv6
prefixes to a allow or deny list privacy extension filter.

The code will create temporary IPv6 addresses from prefixes that
are advertised by Router Advertisement messages. The temporary
IPv6 addresses are then expired according to lifetime and removed.

Fixes #9349

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add information about IPv6 privacy extension

When executing "net iface" command, print current status of
IPv6 privacy extension if it is enabled in config file.

The "net ipv6 ..." command prints IPv6 privacy extension
information, and can add or delete IPv6 prefix filters.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: ipv6: Add privacy extension tests

Add tests that will make sure IPv6 privacy extension code works
as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: ipv6: Select proper IPv6 source address for privacy extension

If IPv6 privacy extension is enabled, then we need to select
proper public or temporary IPv6 source address when sending the
packet.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: ipv6: Add Privacy Extension events to net_mgmt

Add events to be generated when privacy extension is enabled
or disabled, or when allow/deny filter is added or deleted.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Add ref counting to net_if_addr struct

Add reference counting to network interface address (for both
IPv4 and IPv6) so that the address is not removed if there are
sockets using it. If the interface address is removed while there
are sockets using it, the connectivity will fail for the said
socket.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: udp: Add address reference counting tests

Make sure that address reference counting works as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Print ref count information for IPv4 and IPv6 addresses

When printing IPv4 and IPv6 address information in "net ipv4" and
"net ipv6" commands, print also reference count of the IP address.
This is useful for debugging purposes.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Add IPv6 source address selection support

Add support for IPv6 Source Address Selection defined in RFC 5014.
This commit does not add any address selection support to getaddrinfo()
as described in the RFC. The AI_EXTFLAGS symbol and also the ai_eflags
to hints is added but they are not used by the code.
The public/temporary address selection is only done internally by
the net_if_ipv6_select_src_addr_hint() according to flags set by
setsockopt().

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: ipv6: Add address selection tests

Make sure the IPv6 source address selection works as expected
when using the IPV6_ADDR_PREFERENCES socket option.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: ipv6: Generate an event for deprecated address

If IPv6 address expires and moves to deprecated state, then
send a mgmt event so that listeners can act on it.
One such action is to close a socket bound to that source address
so that the address can be released.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo-client: Prefer temporary IPv6 addresses

As a client it makes sense to prefer temporary IPv6 addresses
so mark the socket so.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo-client: Check IPv6 address deprecation

If the IPv6 address that our socket is bound to deprecates, then
re-create the socket so that we avoid using deprecated IPv6 address.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo-client: Avoid flooding output

Do not print progress report so that makes debugging things
a lot harder. It is still possible to enable printouts if needed
by setting PRINT_PROGRESS symbol.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo-server: Prefer public IPv6 addresses

As a server it makes sense to prefer public IPv6 addresses
so mark the socket so.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: context: Add helpers for figuring out local endpoint address

We usually cannot use context->local for established TCP connections
because the local address is not updated for TCP if we are bound to
any address. So create helper that try to figure out the end point
addresses.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: sockets: Fix getsockname() for TCP

The local address used for established TCP socket needs to be
dig out from TCP internals and not from context->local.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: if: Remove IPv6 address when it expires

No need to keep the expired addresses around, just remove it.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/ip/net_context.c:
```diff
@@ -3181,6 +3181,51 @@ int net_context_get_option(struct net_context *context,
+
+	if (IS_ENABLED(CONFIG_NET_TCP) &&
+	    net_context_get_type(ctx) == SOCK_STREAM) {
+		return net_tcp_endpoint_copy(ctx, addr, NULL, addrlen);
```
> We usually cannot use context->local for established TCP connections because the local address is not updated for TCP.

Wouldn't that indicate some sort of bug in our network stack? I think one could expect to be able to access `context->local` address reliably, regardless of the net_context type.

jukkar replied:
I can investigate this more, it is just that if the application does not bind to some specific address, the `context->local` is set to `::`. We do not update the value in this case when the connection is established, don't exactly remember the reasoning for that, but faintly recall there was some issues in it. This part of the code is quite complex and not easy to see all the ramifications.

rlubos reviewed (commented):
That's a lot of work, looks pretty good. Thanks!

jukkar committed:
net: ipv6: Tailor PE values for Zephyr

Making default preferred time much closer to default valid time
so that the served addresses are much shorter time in deprecated
state where they cannot be utilized. This is important in Zephyr
as there are limited number of IPv6 addresses that can be allocated
to the network interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented on include/zephyr/net/net_if.h:
```diff
@@ -279,6 +303,15 @@ struct net_if_ipv6 {
+	 * "DESYNC_FACTOR is a random value within the range 0 - MAX_DESYNC_FACTOR.
+	 * It is computed every time a temporary address is created.
+	 */
+	uint32_t DESYNC_FACTOR;
```
I took the variable/value directly from RFC, easier to map to that mentally. We can of course change this to lower case if needed.

jukkar commented on subsys/net/ip/Kconfig.ipv6:
```diff
@@ -165,6 +165,61 @@ config NET_IPV6_RA_RDNSS
+
+config NET_IPV6_PE_TEMP_VALID_LIFETIME
+	int "Max lifetime for temporary address (in minutes)"
+	default 1800
```
Noticed that too, will fix.

jukkar commented on include/zephyr/net/net_ip.h:
```diff
@@ -1709,6 +1709,47 @@ static inline uint8_t net_priority2vlan(enum net_priority priority)
+/**
+ * @brief Add IPv6 prefix as a privacy extension filter.
+ *
+ * @details Note that the filters can be either black listing or white listing.
```
Yes, missed couple of places where the terminology from the old RFC was used.

jukkar commented on subsys/net/ip/ipv6_pe.c:
```diff
@@ -0,0 +1,713 @@
+ * that can be changed at runtime if needed as recommended in RFC 8981
+ * chapter 3.6.
+ */
+static uint32_t TEMP_VALID_LIFETIME =
```
Easier to map to terminology in RFC which uses capital letters.

jukkar committed:
tests: net: reuseaddr_reuseport: Fix zassert statements

One should not call a function directly from within zassert()
as it can cause the value of errno to be lost as seen by this
failed test output

Assertion failed at tests/net/socket/reuseaddr_reuseport/src/main.c:151:
  test_bind_success: (zsock_bind(sock, addr, addrlen) not equal to 0)
  bind() failed with error 0
FAIL - test_ipv4_udp_bad_both_not_set in 0.000 seconds

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt commented on include/zephyr/net/net_ip.h:
```diff
@@ -1709,6 +1709,47 @@ static inline uint8_t net_priority2vlan(enum net_priority priority)
+ * @details Note that the filters can be either allow or deny listing.
+ *
+ * @param addr IPv6 prefix
+ * @param is_denylist Tells if this filter for allow or deny listing.
```
```suggestion
 * @param is_denylist Tells if this filter is for allowing or denying listing.
```

pdgendt commented on include/zephyr/net/net_ip.h:
```diff
@@ -1709,6 +1709,47 @@ static inline uint8_t net_priority2vlan(enum net_priority priority)
+/**
+ * @brief Add IPv6 prefix as a privacy extension filter.
+ *
+ * @details Note that the filters can be either allow or deny listing.
```
```suggestion
 * @details Note that the filters can either allow or deny listing.
```

pdgendt commented on include/zephyr/net/net_ip.h:
```diff
@@ -1709,6 +1709,47 @@ static inline uint8_t net_priority2vlan(enum net_priority priority)
  */
 const char *net_family2str(sa_family_t family);
 
+#if defined(CONFIG_NET_IPV6_PE)
```
Do we need to do something special for `doxygen`?

jukkar replied:
Good point, yes we do.

pdgendt commented on subsys/net/ip/ipv6.h:
```diff
@@ -595,5 +595,91 @@ static inline void net_ipv6_set_ecn(uint8_t *tc, uint8_t ecn)
+static inline int net_ipv6_pe_filter_foreach(net_ipv6_pe_filter_cb_t cb,
+					     void *user_data)
+{
+	return 0;
```
```suggestion
	ARG_UNUSED(cb);
	ARG_UNUSED(user_data);

	return 0;
```

pdgendt commented on subsys/net/ip/ipv6_pe.c:
```diff
@@ -0,0 +1,758 @@
+
+	return ret;
+#else
+	return true;
```
```suggestion
	ARG_UNUSED(prefix);

	return true;
```

pdgendt commented on subsys/net/ip/ipv6_pe.c:
```diff
@@ -0,0 +1,758 @@
+
+	return ret;
+#else
+	return -ENOTSUP;
```
```suggestion
	ARG_UNUSED(addr);
	ARG_UNUSED(is_denylist);

	return -ENOTSUP;
```

pdgendt commented on subsys/net/ip/ipv6_pe.c:
```diff
@@ -0,0 +1,758 @@
+
+	return ret;
+#else
+	return -ENOTSUP;
```
```suggestion
	ARG_UNUSED(addr);

	return -ENOTSUP;
```

pdgendt commented on subsys/net/ip/ipv6_pe.c:
```diff
@@ -0,0 +1,758 @@
+
+	return count;
+#else
+	return 0;
```
```suggestion
	ARG_UNUSED(cb);
	ARG_UNUSED(user_data);

	return 0;
```

pdgendt commented on subsys/net/ip/ipv6_pe.c:
```diff
@@ -0,0 +1,758 @@
+
+bool net_ipv6_pe_check_dad(int count)
+{
+	if (count > TEMP_IDGEN_RETRIES) {
+		return false;
+	}
+
+	return true;
```
```suggestion
	return count <= TEMP_IDGEN_RETRIES;
```

pdgendt reviewed (dismissed):
LGTM, just some nits

jukkar committed:
tests: net: coap: Enable loopback interface

There needs to be at least one network interface in order
to avoid this error.

E: There is no network interface to work with!

So enable loopback support which creates a loopback interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: if: ipv6: Do not crash if there are no network interfaces

When trying to get IPv6 prefix, if there are no network interfaces
then the net_if_get_default() will return NULL but we should not
crash at that point but return gracefully.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: if: Print warning if there are no network interfaces

When calling net_if_get_default(), print a warning if no
network interfaces are found. This helps debugging mysterious
crashes.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* fixed according to comments
* more CI fixes
